### PR TITLE
Centralize async callback storage in preparation for IPC/error handling

### PR DESCRIPTION
**Pull request by stevenewald**
This is in preparation for error handling and IPC, which will both call async user callbacks. With the new ProcessCallbackStorage, simply adding a callback (now using etl::delegate) will cause it to eventually be triggered by the scheduler. Nice!

Aside: right now we're using the pattern of many things being a singleton, while another option is to have the scheduler own everything (hierarchy). 

One could argue the latter is better practice because singletons can be messy, but I'd argue most(all?) of the singletons we have right now _must_ be singletons (how could you have two button drivers? two schedulers?).

However, this makes visualizing dependencies (constr/deconstr and general call patterns) much harder.

In reality, it doesn't matter, just interesting to think about.

### CODE REVIEW

File: include/drivers/button_driver.hpp
Comment (echavemann) [inline include/drivers/button_driver.hpp:1]: this is just buttons.hpp yea? not sure why the diff is not a move

File: include/hal/gpio_event_controller.hpp [hunk 1]
```diff
@@ -11,27 +11,23 @@ static constexpr uint8_t GPIO_PINS = 32;
...
-        static GPIOEventController controller;
-        return controller;
-    }
+    static GPIOEventController& get();
 
-    GPIOEventController& operator=(const GPIOEventController&) = delete;
-    GPIOEventController& operator=(GPIOEventController&&) = delete;
-    GPIOEventController(const GPIOEventController&) = delete;
-    GPIOEventController(GPIOEventController&&) = delete;
+    // GPIOEventController& operator=(const GPIOEventController&) = delete;
```
Comment (echavemann) [inline include/hal/gpio_event_controller.hpp:27]: can we del if not needed?

File: include/process_callback_storage.hpp [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+#include "config.hpp"
+#include "drivers/driver_enums.hpp"
+
+namespace edge {
+// One option is to make the drivers and other things non singletons and have the
+// scheduler own everything. I'm actually okay with this, but not a big deal rn and this
+// works fine
+// This doesn't matter rn but would matter if we were worried about shutdown and wanted
+// very predictable dtor calls
+// TLDR: singleton vs everything owned by scheduler, food for thought
```
Comment (echavemann) [inline include/process_callback_storage.hpp:11]: in favor of the latter i think

Comment (echavemann) [reply include/process_callback_storage.hpp:11]: but yea functionally pretty similar

**Review by echavemann — APPROVED** [review]

File: include/hal/gpio_event_controller.hpp (hunk 1 above)
Comment (stevenewald) [reply include/hal/gpio_event_controller.hpp:27]: whoops accidentally commented out

**Pull request merged**